### ISSUE TITLE
The environment variable set in sendRequest (prerequest script) during Newman execution not accessible in testscript. Works in Collection Runner and Postman.

### ISSUE THREAD

**Issue by slavenantic** [postmanlabs/newman#1825]
1. Newman Version (can be found via `newman -v`): 4.2.2
 2. OS details (type, version, and architecture): Windows 10 Pro, 64
 3. Are you using Newman as a library, or via the CLI? CLI
 3. Did you encounter this recently, or has this bug always been there: Recently
 4. Expected behaviour: To update the environment variable inside the sendRequest in the same manner as from Collection Runner and Postman
 5. Command / script used to run Newman: (pre-request script)
`// 
let id1 = 19;
let isDataDriven = pm.iterationData.get("Input");
let id2 = 0;

if (isDataDriven) {
    id1 = pm.iterationData.get("Input").id1;
    id2 = pm.iterationData.get("Input").id2;
    pm.environment.set('id2', id2);
}

var requestUrl = pm.environment.get("environment") + '/resource/' + id1;
var options = { method: 'GET', url: requestUrl };

pm.sendRequest(options, function (error, response) {
    if (error) throw new Error(error);
    var jsonData = response.json();
    pm.environment.set('responseTime', response.responseTime);
    if (id2 !== 0) {
        id2 = jsonData.id2;
        pm.environment.set('id2', id2);
    }
});`
 6. Sample collection, and auxiliary files (minus the sensitive details):
 7. Screenshots (if applicable):

**Comment by slavenantic** [comment]
Is there any chance to send these things to you via email?

**Comment by codenirvana** [comment]
@slavenantic You can attach the collection JSON file(minus the sensitive details) in GitHub comment itself.
For that collection, don't use your private endpoint, you can use any sample or echo domain.

**Comment by slavenantic** [comment]
Here they go.

[sample.postman_collection.zip](https://github.com/postmanlabs/newman/files/2693478/sample.postman_collection.zip)

**Comment by slavenantic** [comment]
There you go:
newman run "C:\PathToCollection\laplace.postman_collection.json"   --folder "3_4_1_CheckRecipesWithImagePerformance" -e "C:\PathToEnvironment\qa.postman_environment.json" -d "C:\PathToFile\allImageIds_nl.json" -n 100 --reporters cli  --disable-unicode > "C:\PathToReports\3_Images\3_4_PerformanceTests\3_4_1_CheckRecipesWithImagePerformance\cli\20190103135332\allImageIds_nl.txt"****

**Comment by codenirvana** [comment]
@slavenantic Can you try execution the attached collection using:
```console
$ newman run NEWMAN-1825.postman_collection.json -d allImageIds_nl.json
```
Modified your scripts to reproduce the issue locally. The environment variable `recipeResponseTime` is set in `sendRequest` correctly.

[NEWMAN-1825.zip](https://github.com/postmanlabs/newman/files/2723798/NEWMAN-1825.zip)

**Comment by slavenantic** [comment]
I can still reproduce the same issue with your modifications (I don't see any relevant changes in your collection). The variable is still set at the same place in pre-request script. What changes did you make actually?

**Comment by codenirvana** [comment]
@slavenantic I just changed the request endpoint to an echo server to execute the collection.

Are tests failing for you? Can you try updating Newman to the latest version and check if the issue still persists?

**Comment by slavenantic** [comment]
I am still having the same issue, even after newman update to 4.3.1.

**Comment by codenirvana** [comment]
@slavenantic As discussed above, I am not able to reproduce this issue.
Check the attached collection in which a variable set in sendRequest(prerequest) is accessible in the test script.

[NEWMAN-1825.json.zip](https://github.com/postmanlabs/newman/files/2763013/NEWMAN-1825.json.zip)

**Comment by praveendvd** [comment]
@codenirvana I am facing the same issue,

I have uploaded the collection file, it has two requests one that works, one that doesn't.

I want to update the collection variable (or any variable) according to a few criteria and the status of the request , 

[collection.zip](https://github.com/postmanlabs/newman/files/4301810/collection.zip)

**Comment by csk1827** [comment]
@praveendvd you are updating the "projectid" variable in a **promise** in the test script of the second request, which is taking some time to execute.  **As promises are executed asynchronously**, the getter of the "projectid" variable in (the last line of the test script) is getting the old value of the variable, before the execution of the promise. That's the reason why you are getting the old value for the "projectid" variable.

**Comment by csk1827** [comment]
@codenirvana It might be the same case with @slavenantic too. 

I have a question for you. Are the Postman sandboxes **strictly sequential** or they can run in **parallel** also? If they run in parallel, then some of the variables may not get assigned/updated in time(before the next sandbox tries to access that variable) if a request sent by sendRequest() in a sandbox takes a long time to execute.

**Comment by praveendvd** [comment]
@csk1827 yes that's the issue I was able to fix it
```

setTimeout(function() {

       pm.test("Status code is 200", function () {
              pm.response.to.have.status(200);
        });

       pm.test("Status code is 200", function () {
          pm.response.to.have.status(200);
       });

}, 5000);
```,

**Comment by shamasis** [comment]
Sandboxes are supposed to be executed serially. The way sandbox waits for script execution completion is to look for timers to end. Since promises are not timers, they aren’t waited for. To wait for it, we need to deprecate till node v6 (and maybe 8) and use async_hooks to wait for timers. cc @codenirvana @coditva

**Comment by codenirvana** [comment]
Closing as duplicate of https://github.com/postmanlabs/postman-app-support/issues/5606
